content/2023-10-02-W-global-constructors.rst

```diff
@@ -0,0 +1,162 @@
+-Wglobal-constructors
+#####################
+
+:date: 2023-10-02-23
+:category: mozilla
+:lang: en
+:authors: Serge Guelton
+:summary: What's worse than a global variable? A global variable initialized at
+          startup!
+
+
+.. code-block:: c++
+
+    #include <iostream>
+    #include <unordered_set>
+    #include <string>
+
+    const std::unordered_set<std::string> primary_colors = {"red", "green", "blue"};
+
+    int main(int argc, char const ** argv) {
+        char const * color = argc > 1 ? argv[1] : "pink";
+        return primary_colors.count(color) ? 0 : 1;
+    }
+
+Let's compile and run the above snippet:
+
+.. code-block:: shell
+
+   $ clang++ --version
+   clang version 14.0.5 (Fedora 14.0.5-2.fc36)
+   Target: x86_64-redhat-linux-gnu
+   Thread model: posix
+   InstalledDir: /usr/bin
+
+   $ clang++ colors.cpp -O2 -o colors && ./colors green && echo ok
+   ok
+
+Let's be a bit more pedantic and turn all warnings on:
+
+
+.. code-block:: shell
+
+   $ clang++ colors.cpp -Wall -O2 -o colors && ./colors green && echo ok
+   ok
+
+No, I meant really pendantic
+
+.. code-block:: shell
+
+   $ clang++ colors.cpp -Weverything -O2 -o colors && ./colors green && echo ok
+    colors.cpp:5:56: warning: initialization of initializer_list object is incompatible with C++98 [-Wc++98-compat]
+    const std::unordered_set<std::string> primary_colors = {"red", "green", "blue"};
+                                                           ^~~~~~~~~~~~~~~~~~~~~~~~
+    colors.cpp:5:39: warning: declaration requires an exit-time destructor [-Wexit-time-destructors]
+    const std::unordered_set<std::string> primary_colors = {"red", "green", "blue"};
+                                          ^
+    colors.cpp:5:39: warning: declaration requires a global destructor [-Wglobal-constructors]
+    3 warnings generated.
+    ok
+
+
+Now that's what I needed for my intro: ``-Wglobal-constructors`` (let's forget
+about C++98, right?). ``-Wexit-time-destructors`` is a subset of
+``-Wglobal-constructors``, so let's focus on global constructors.
+
+Indeed we are creating a ``const`` unordered set named ``colors``, but the
+initialization code has to happen at some point. None of `its constructors
+<https://en.cppreference.com/w/cpp/container/unordered_set/unordered_set>`_ is
+flagged as ``constexpr`` so the initialization happens at startup. Before the
+call to the ``main`` function.
+
+Why is it worth a warning? The `clang manual
+<https://clang.llvm.org/docs/DiagnosticsReference.html#wglobal-constructors>`_
+is not very talkative on the subject, and the `commit introducing the feature
+<https://github.com/llvm/llvm-project/commit/47e40931c9af037ceae73ecab7db739a34160a0e>`_
+is not super helpful either. But we can guess the reasons: bug fixing (think `Static Initialization
+Order Fiasco <Static Initialization Order Fiasco>`_) and performance
+(increased startup time).
```

Comment:
in the conclusion, maybe you could add a link pointing to your blog post 
and/or a commit improving the doc ;)